lib/rust/performance-logger/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+        self.start_time + self.duration
+    }
+}
+
+impl Display for Measurement {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str(&format!(
+            "[{:.2}, {:.2}] ({:.2}) {}",
+            self.start_time,
+            self.end_time(),
+            self.duration,
+            self.name
+        ))
+    }
+}
+
+/// Error that can occur when converting between a `PerformanceEntry` and a `Measurement`.
+#[derive(Debug, Clone)]
+pub enum MeasurementConversionError {
+    /// The name of the `PerformanceEntry` was not in a format that could be decoded to a
+    /// `Measurement`.
+    InvalidFormatting,
+    /// The log level encoded in the `PerformanceEntry`s name is not a valid log level.
+    InvalidLogLevel,
+}
+
+impl TryFrom<PerformanceEntry> for Measurement {
+    type Error = MeasurementConversionError;
+
+    fn try_from(measure: PerformanceEntry) -> Result<Self, Self::Error> {
+        use MeasurementConversionError::*;
+
+        let start_time = measure.start_time();
+        let duration = measure.duration();
+        let name_js = measure.name();
+        let name_parts: Vec<_> = name_js.split(MESSAGE_DELIMITER).collect();
+        if name_parts.len() < 2 {
+            return Err(InvalidFormatting); //FIXME
+        }
+        let name = name_parts[1].to_string();
+        let log_level_name = name_parts[0].to_string();
+
+        let log_level: LogLevel =
+            serde_plain::from_str(&log_level_name).or(Err(InvalidLogLevel))?;
+        Ok(Measurement { start_time, duration, name, log_level })
+    }
+}
+
+/// Error that can occur when taking a measurement.
+#[derive(Clone, Debug)]
+pub enum MeasurementError {
+    /// Parsing the measurement information from the performance API failed, for example, due to an
+    /// invalid log level.
+    InvalidMeasurementConversion,
+    /// No measurement was created in the performance API backend.
+    NoMeasurementFound,
+    /// A function call to the Performance API failed to execute.
+    PerformanceAPIExecutionFailure,
+    /// Profiling for the given profiling level was disabled.
+    ProfilingDisabled,
+}
+
+
+
+// =================
+// === Log Level ===
+// =================
+
+#[derive(Copy, Clone, Debug)]
+#[derive(Serialize, Deserialize)]
+#[allow(missing_docs)]
+pub enum LogLevel {
+    SECTION,
+    TASK,
+    DETAIL,
+    DEBUG,
+}
+
+
+
+// =========================
+// === Profiling Methods ===
+// =========================
+
+/// Delimiter used to to encode information in the `PerformanceEntry` name.
+const MESSAGE_DELIMITER: &str = "//";
+
+/// Check at compile time whether the given log level should perform any logging activity.
+const fn log_level_is_active(log_level: LogLevel) -> bool {
+    match log_level {
+        LogLevel::SECTION => ENABLE_SECTION_PROFILING,
+        LogLevel::TASK => ENABLE_TASK_PROFILING,
+        LogLevel::DETAIL => ENABLE_DETAIL_PROFILING,
+        LogLevel::DEBUG => ENABLE_DEBUG_PROFILING,
+    }
+}
```

Comment:
code repetition. Adding a new level should b e as automated as possible. Such boilerplate code should not exist.

Reply:
Refactored.

Reply:
This is the only repetition that I left. It seems alternatives (like the registry in JS) are actually more complex and uglier. This will never be missed when adding a new `LogLevel` as the compiler will warn about the match statement and the change required will be obvious to the developer.